harness/determined/core/_context.py

```diff
@@ -53,6 +53,8 @@ def __enter__(self) -> "Context":
     def __exit__(self, typ: type, value: Exception, tb: Any) -> None:
         self.preempt.close()
         self.distributed.close()
+        self.train.close()
+        self.checkpoint.close()
```

Comment:
The object hierarchy here is a bit unclear.

Does the train context own the tensorboard manager?  Does the checkpoint context own the tensorboard manager?  Either could close it, but what happens if the trial context is closed while the checkpoint context still intends to use it?

To answer the last part, in our system, is that the train context and the checkpoint context are not ever distinct entities that one might be operating after the other is done.  This is guaranteed by the fact that core.Context is the object we give to users, and core.Context is owner of both and core.Context manages their lifetimes together.

It stands to reason that if both the checkpoint context and train context need this tensorboard manager, perhaps the parent object core.Context should also own the tensorboard manager, and pass it in as a parameter to train context and checkpoint context.  That way, the ownership is clearer:
- core.Context owns the tensorboard manager and is responsible for closing it at the right time
- train context uses the tensorboard manager, and trusts that its owner manages the lifetime of the tensorboard manager properly (train context assumes the tensorboard manager is always usable throughout the train context lifetime)
- checkpoint context is the same story as the train context

This is actually the exact same ownership pattern as the distributed context, which is handed as a parameter to several other context objects.  But only the core.Context closes the distributed context.

I like single ownership patterns because they're easy to understand and they're usually flexible enough.  More complex patterns exist (reference counted patterns, for instance) but those are best left to when they're strictly necessary.